Support '-l' option

blacktoad30 opened the pull request:
* 04.ls/lib/ls_methods.rb (#table_print): Remove method.
(#adjust_list, #print_table, #file_info, #file_mode, #entry_type)
(#entry_perm, #file_modified_six_months_ago?): New methods.

* 04.ls/ls.rb: Get '-l' option from ARGV.

* 04.ls/test/check-ruby-ls.sh: New File.

blacktoad30 committed:
Improve processing in '#matrix'

* 04.ls/lib/ls_methods.rb (#matrix): Use 'Array#concat' and 'Array#new'
instead of 'Integer#times' and 'Array#push'.

blacktoad30 committed:
Support '-l' option

* 04.ls/lib/ls_methods.rb (#table_print): Remove method.
(#adjust_list, #print_table, #file_info, #file_mode, #entry_type)
(#entry_perm, #file_modified_six_months_ago?): New methods.

* 04.ls/ls.rb: Get '-l' option from ARGV.

* 04.ls/test/check-ruby-ls.sh: New File.

maedana commented on 04.ls/lib/ls_methods.rb:
```diff
@@ -10,31 +12,109 @@ def dot_file?(fname)
+        elm.to_s.ljust(width, padding)
+      when :right
+        elm.to_s.rjust(width, padding)
+      when :center
```
centerの利用パターンなさそうに見えますね。不要なのでは?

maedana commented on 04.ls/lib/ls_methods.rb:
```diff
@@ -10,31 +12,109 @@ def dot_file?(fname)
   ary.each_slice(col).to_a
 end
 
+def adjust_list(list, padding = ' ', align: :left, suffix: '')
```
paddingで他の値を指定するパターンがなさそうですし、メソッド内ローカル変数等にして引数からは無くしたらいいんじゃないかな、と思いました。

maedana commented on 04.ls/ls.rb:
```diff
@@ -2,5 +2,62 @@
+           .transpose
+else
+  tbl = matrix(entries, 3)
+  table =
```
tblってtableの略ですよね。つまり違うものを同じ変数名にしてるようなものだと思うんです。メソッド名もmatrixですしメソッド名と全くことなる命名の変数名にするのは基本的に可読性下げてると思うんです。いかがでしょうか?

maedana commented on 04.ls/ls.rb:
```diff
@@ -2,5 +2,62 @@
+  tbl = matrix(entries, 3)
+  table =
+    tbl.map
+       .with_index do |lst, idx|
```
以前からある傾向だと思いますが、微妙に省略したりしますよね。lstとlistで一文字しか違わないですし変に省略しないほうがいいような気がします。添字のidxはむしろ`i`だけにするのが一般的なんじゃないですかね。

そもそもtbl、つまりmatrixの1つ1つの命名としてlist(lst)は適切なのだろうか?というのも気になりますね。

そこまで行数の多くないブロックのブロック変数ですし、生存期間も短いのでそこまで目くじら立てなくてもな、とは思うのですけども。

配列を返すメソッドはそれを利用するときにどんな変数に入れるのか、ループするとしたら一つ一つの要素をどう表現するのか、みたいなことを考えて命名するとスッキリすることが多いかなと思いました。

```ruby
entries = child_files('.')
...
entries.map { |fname|
```
みたいなのもそうですね。同じものを示しているはずなのにentry, child_file, fnameという異なる3つの単語が出てきていて可読性下げてると思うんです。いかがでしょうか。

maedana commented on 04.ls/ls.rb:
```diff
@@ -2,5 +2,62 @@
+           .map
+           .with_index do |lst, idx|
+             case idx
+             when 0, 2, 3
```
このへんの添字の値はマジックナンバーなので、定数にするなどしてリーダブルにしたほうがいいのではないですかね。例えば1年後にメンテナンスすることを想像したときに3て何を示してるんだっけ... とならないですか?

maedana commented on 04.ls/ls.rb:
```diff
@@ -2,5 +2,62 @@
+
+if opts['l']
+  blocks =
+    entries.map do |fname|
```
この時点で sum使ってしまってもいいかなーと思います。ブロックを渡す使い方ですね。
https://docs.ruby-lang.org/ja/latest/method/Array/i/sum.html

maedana commented on 04.ls/lib/ls_methods.rb:
```diff
@@ -10,31 +12,109 @@ def dot_file?(fname)
   str.to_s.length + str.to_s.grapheme_clusters.count { |c| !c.ascii_only? }
 end
+
+def print_table(table, sep: ' ')
```
同上ですがsepに他の値を渡すパターンがなさそうなので引数でなくてもいいのかな、と思いました。

maedana commented on 04.ls/lib/ls_methods.rb:
```diff
@@ -10,31 +12,109 @@ def dot_file?(fname)
+
+MOD_X = [%w[- x S s], %w[- x S s], %w[- x T t]].freeze
+
+def entry_perm(fmode)
```
permっていう略し方は一般的なのですかね? 普通にpermissionと書いたほうがわかりやすい気がしました。

blacktoad30 committed:
Remove unused branch and keyword arguments

* 04.ls/lib/ls_methods.rb (#adjust_list): Remove ':center' case branch.
(#adjust_list, #print_table): Remove unused keyword arguments.

blacktoad30 committed:
Inline redundant methods

* 04.ls/lib/ls_methods.rb (#dot_file?, #max_length): Remove method.
(#child_files): Use 'String#match?' instead of '#dot_file?'.
(#adjust_list): Inline '#max_length'.

blacktoad30 committed:
Extract code blocks into separate methods

* 04.ls/lib/ls_methods.rb (#main, #tabulate_file_names, #total_blocks)
(#file_infos): New method extracted from top level.
(#file_info): Remove unnecessary temporary variables.
(#file_modified_six_months_ago?): Remove method.
(#file_rdev_or_size, #file_modified_date_time)
(#file_path_name): New method extracted from '#file_info'.
(TIME_NOW, SIX_MONTHS_AGO): New constant.

* 04.ls/ls.rb: Just call '#main' with argument 'ARGV'.

blacktoad30 committed:
Avoid using magic numbers in conditional expressions

* 04.ls/lib/ls_methods.rb (FileInfo): New subclass.
(#file_infos, #file_info): Use 'FileInfo'.

blacktoad30 committed:
Improve style of methods

* 04.ls/lib/ls_methods.rb (#child_files): Chain 'Array#sort' at last.
(#tabulate_file_names, #matrix, #entry_perm): Change chaining method.
(#total_blocks): Simplify method chaining.

blacktoad30 committed:
Rename methods and constant

* 04.ls/lib/ls_methods.rb (#matrix, #entry_type, MOD_X, #entry_perm):
Rename to ...
(#split_list_into_rows, #file_type_char, FILE_MODE_EXEC)
(#file_permission): ... them.
(#tabulate_file_names, #file_mode): Change due to rename above.

blacktoad30 committed:
Rename parameters

* 04.ls/lib/ls_methods.rb (#child_files, #tabulate_file_names)
(#split_list_into_rows, #total_blocks, #file_infos, #file_info)
(#file_mode, #file_type_char, #file_permission): Rename method
parameter.
(#file_info, #file_mode, #file_permission): Rename local variable.
(#child_files, #print_table, #file_infos, #file_permission): Rename
block parameter.

blacktoad30 committed:
Fix test script

* 04.ls/test/check-ruby-ls.sh ($cmd_ls) <max_width>: Change argument
to '$ext_width'.

blacktoad30 commented:
@maedana ご無沙汰しております。返信が遅くなってしまい、申し訳ございません。
ご指摘いただいていた箇所について、修正いたしました。
また、修正作業中に気付いた部分 (メインでの処理をメソッド化する、等) についても修正しております。
お手数ですが、ご確認のほどよろしくお願いいたします。

maedana commented on 04.ls/lib/ls_methods.rb:
```diff
@@ -1,40 +1,173 @@
+  files.map(&File.method(:lstat)).sum(&:blocks).ceildiv(2)
+end
+
+FileInfo = Data.define(:mode,
```
全然知らなかったです。Ruby3.2から使えるんですねー。イミュータブルなデータの入れ物クラスを作りたいときに便利そうですね。仕事で使ってみようと思いました :+1: 

特に根拠があるわけじゃないのですけど定義場所はファイル冒頭ぐらいがいい気がしました。

maedana commented on 04.ls/lib/ls_methods.rb:
```diff
@@ -1,40 +1,173 @@
+  infos = files.map { |file| file_info(file) }
+
+  table =
+    Enumerator.new do |y|
```
なぜ`y`?と思ったら、Enumerator::Yielderなんですね。 https://docs.ruby-lang.org/ja/latest/method/Enumerator/s/new.html

maedana commented on 04.ls/lib/ls_methods.rb:
```diff
@@ -1,40 +1,173 @@
+  end
+end
+
+FILE_MODE_EXEC = [%w[- x S s], %w[- x S s], %w[- x T t]].freeze
```
細かいですが、定数はファイル上部にまとめて定義するのかセオリーかなと思いました。

maedana commented on 04.ls/lib/ls_methods.rb:
```diff
@@ -1,40 +1,173 @@
+end
+
+def total_blocks(files)
+  files.map(&File.method(:lstat)).sum(&:blocks).ceildiv(2)
```
ちょっとトリッキーに感じました。普通に`files.sum { |f| File.lstat(f).blocks }.ceildiv(2)`とかでいいのではないでしょうか。

maedana commented on 04.ls/lib/ls_methods.rb:
```diff
@@ -1,40 +1,173 @@
-def max_length(str_list)
-  str_list.map { |str| monofont_width(str) }.max
+def adjust_list(list, align: :left, suffix: '')
+  width = list.map { |elm| monofont_width(elm.to_s) }.max
```
他で書いていたかもしれないですが、Rubyにおいては配列の命名は名詞の複数形がわかりやすいかなーと思います。`list`には配列に印象ありますけど、ブロックを伴う処理をするときにブロック変数の命名苦しくなりませんか? ここも`list`に対して`elm`になってますよね。適切な命名かどうかはさておき、`elements.map { |element| ... }`となってるほうが素直で読みやすいかなーと思います。

maedana commented on 04.ls/lib/ls_methods.rb:
```diff
@@ -1,40 +1,173 @@
+  table =
+    Enumerator.new do |y|
+      FileInfo.members.each do |info_type|
+        col = infos.map(&info_type)
```
`col`だと配列だとわかりにくいです。複数形の命名にしたいですね
